slo_lambda/src/windowed_slo.py

```diff
@@ -1,203 +1,258 @@
+        return total * self.multiplier
 
+    def extract_stat(self, datapoint: Dict) -> float:
+        """
+        extract_stat takes a Cloudwatch datapoint and returns the numeric value
+        """
+        if self.extended_statistic:
+            return datapoint["ExtendedStatistics"][self.extended_statistic]
+
+        return datapoint[self.statistic]
 
-class BackendSuccessMetric:
+
+class CompositeMetric:
     """
-    Metric that combines a few others 
+    Holds multiple SingleMetrics allowing for returning the sum of the metrics.
     """
-    def sum(self, client) -> float:
-        request_count = METRICS['request_count'].sum(client)
-        elb_500s = METRICS['elb_500s'].sum(client)
-        target_500s = METRICS['target_500s'].sum(client)
 
-        return request_count - target_500s
+    def __init__(self, window_days: int, metrics: List[Dict]):
+        self.metrics = [SingleMetric(window_days=window_days, **m) for m in metrics]
 
+    def sum(self) -> float:
+        total = 0.0
 
-class ValidRequests:
-    """
-    Metric that encompasses all valid requests to the LB, including those that
-    return before selecting a target group. Excludes 4XX reponses originiating
-    from the LB.
-    """
-    def sum(self, client) -> float:
-        request_count = METRICS['request_count'].sum(client)
-        elb_500s = METRICS['elb_500s'].sum(client)
+        for m in self.metrics:
+            total += m.sum()
+
+        return total
 
-        return request_count + elb_500s
 
-class GoodResponses:
-    """
-    Metric that encompasses all good reponses from the target group. Excludes
-    5XX responses and timeouts.
+class SLI:
     """
-    def sum(self, client) -> float:
-        responses = [
-            METRICS['target_200s'],
-            METRICS['target_300s'],
-            METRICS['target_400s'],
-        ]
-        return sum([x.sum(client) for x in responses])
+    SLI calculates an SLI definition as a ratio of good requests to
+    total requests.
 
+    Please note that we assume `numerator` and `denominator` were defined in
+    Terraform, and parsed as JSON. So they lists of Dicts (for composite metrics).
 
-class SLI:
-    def __init__(self, name: str, num: Metric, denom: Metric):
-        self.name = name
-        self.num = num
-        self.denom = denom
+    Example:
 
-    def get_ratio(self, client) -> float:
-        """
-        Can return ZeroDivisonError. Make sure to catch it.
+    interesting_availability = {
+      window_days = 30
+      numerator = {
+        namespace   = "foo_env/sli"
+        metric_name = "InterestingUrisSuccess"
+        dimensions = [
+          {
+            Name  = "Hostname"
+            Value = "foo"
+          }
+        ]
+      }
+      ...
+    }
+    """
+
+    def __init__(
+        self,
+        numerator: List[Dict],
+        denominator: List[Dict],
+        window_days: int = WINDOW_DAYS,
+    ):
+        if window_days is None:
```

Comment:
You set the default already if we didn't pass in `window_days`, so this teems redundant.

Reply:
Unfortunately, Terraform sets this to `None` if `window_days` isn't explicitly set there, so we need to take care of that case as well.

Reply:
Ah, I see!  OK.